File: scripts/metrics_collector.py
"""
Metrics Collector for Federated Learning Experiments
Collects and saves all experimental metrics to JSON/CSV for thesis analysis.
"""

import json
import csv
import os
from datetime import datetime
from typing import Dict, List, Any, Optional
from pathlib import Path
import logging

logger = logging.getLogger(__name__)


class MetricsCollector:
    """Collects and stores experimental metrics"""
    
    def __init__(self, experiment_id: str, results_dir: str = "results"):
        """
        Initialize metrics collector.
        
        Args:
            experiment_id: Unique identifier for this experiment
            results_dir: Directory to save results
        """
        self.experiment_id = experiment_id
        self.results_dir = Path(results_dir)
        self.results_dir.mkdir(parents=True, exist_ok=True)
        
        # Store all collected metrics
        self.experiment_data = {
            "experiment_id": experiment_id,
            "timestamp": datetime.now().isoformat(),
            "config": {},
            "rounds": [],
            "final_metrics": {},
            "client_metrics": [],
            "mobile_metrics": []
        }
        
        logger.info(f"Initialized metrics collector for experiment: {experiment_id}")
    
    def set_config(self, config: Dict[str, Any]):
        """
        Set experiment configuration.
        
        Args:
            config: Dictionary with experiment parameters
        """
        self.experiment_data["config"] = config
        logger.info(f"Set experiment config: {config}")
    
    def add_round_metrics(
        self,
        round_num: int,
        train_loss: float,
        test_metrics: Optional[Dict[str, float]] = None,
        aggregation_info: Optional[Dict[str, Any]] = None,
        client_metrics: Optional[List[Dict[str, Any]]] = None
    ):
        """
        Add metrics for a training round.
        
        Args:
            round_num: Round number
            train_loss: Average training loss
            test_metrics: Test set evaluation metrics (NDCG, Hit Rate, etc.)
            aggregation_info: Aggregation statistics
            client_metrics: Per-client metrics
        """
        round_data = {
            "round": round_num,
            "train_loss": train_loss,
            "test_metrics": test_metrics or {},
            "aggregation": aggregation_info or {},
            "client_metrics": client_metrics or [],
            "timestamp": datetime.now().isoformat()
        }
        
        self.experiment_data["rounds"].append(round_data)
        logger.info(f"Added metrics for round {round_num}")
    
    def add_final_metrics(self, metrics: Dict[str, Any]):
        """
        Add final evaluation metrics.
        
        Args:
            metrics: Final metrics dictionary
        """
        self.experiment_data["final_metrics"] = metrics
        logger.info(f"Added final metrics: {metrics}")
    
    def add_client_metrics(self, client_id: str, metrics: Dict[str, Any]):
        """
        Add per-client metrics.
        
        Args:
            client_id: Client identifier
            metrics: Client-specific metrics
        """
        client_data = {
            "client_id": client_id,
            **metrics,
            "timestamp": datetime.now().isoformat()
        }
        self.experiment_data["client_metrics"].append(client_data)
    
    def add_mobile_metrics(self, device_id: str, metrics: Dict[str, Any]):
        """
        Add mobile device resource metrics.
        
        Args:
            device_id: Device identifier
            metrics: Resource metrics (battery, memory, CPU, etc.)
        """
        mobile_data = {
            "device_id": device_id,
            **metrics,
            "timestamp": datetime.now().isoformat()
        }
        self.experiment_data["mobile_metrics"].append(mobile_data)
    
    def save_json(self, filename: Optional[str] = None) -> str:
        """
        Save metrics to JSON file.
        
        Args:
            filename: Optional custom filename (default: experiment_id.json)
        
        Returns:
            Path to saved file
        """
        if filename is None:
            filename = f"{self.experiment_id}.json"
        
        filepath = self.results_dir / filename
        
        with open(filepath, 'w') as f:
            json.dump(self.experiment_data, f, indent=2)
        
        logger.info(f"Saved metrics to {filepath}")
        return str(filepath)
    
    def save_csv_summary(self, filename: Optional[str] = None) -> str:
        """
        Save summary to CSV file (useful for quick analysis).
        
        Args:
            filename: Optional custom filename
        
        Returns:
            Path to saved file
        """
        if filename is None:
            filename = f"{self.experiment_id}_summary.csv"
        
        filepath = self.results_dir / filename
        
        # Create summary CSV with key metrics per round
        with open(filepath, 'w', newline='') as f:
            writer = csv.writer(f)
            
            # Header
            writer.writerow([
                'Round', 'Train_Loss', 'NDCG@10', 'Hit@10', 'Precision@10', 
                'Recall@10', 'MSE', 'MAE', 'Accuracy', 'Num_Clients', 'Total_Samples'
            ])
            
            # Round data
            for round_data in self.experiment_data["rounds"]:
                test_metrics = round_data.get("test_metrics", {})
                agg = round_data.get("aggregation", {})
                
                writer.writerow([
                    round_data["round"],
                    round_data.get("train_loss", ""),
                    test_metrics.get("NDCG@10", ""),
                    test_metrics.get("Hit@10", ""),
                    test_metrics.get("Precision@10", ""),
                    test_metrics.get("Recall@10", ""),
                    test_metrics.get("mse", ""),
                    test_metrics.get("mae", ""),
                    test_metrics.get("accuracy", ""),
                    agg.get("num_clients", ""),
                    agg.get("total_samples", "")
                ])
            
            # Final metrics row
            final = self.experiment_data.get("final_metrics", {})
            writer.writerow([
                "FINAL",
                "",
                final.get("NDCG@10", ""),
                final.get("Hit@10", ""),
                final.get("Precision@10", ""),
                final.get("Recall@10", ""),
                final.get("mse", ""),
                final.get("mae", ""),
                final.get("accuracy", ""),
                "",
                ""
            ])
        
        logger.info(f"Saved CSV summary to {filepath}")
        return str(filepath)
    
    def get_summary(self) -> Dict[str, Any]:
        """
        Get summary statistics of collected metrics.
        
        Returns:
            Summary dictionary
        """
        rounds = self.experiment_data["rounds"]
        
        if not rounds:
            return {"message": "No rounds collected yet"}
        
        # Extract metrics across rounds
        train_losses = [r["train_loss"] for r in rounds if "train_loss" in r]
        ndcg_scores = [r["test_metrics"].get("NDCG@10", 0) for r in rounds if "test_metrics" in r]
        hit_scores = [r["test_metrics"].get("Hit@10", 0) for r in rounds if "test_metrics" in r]
        
        return {
            "experiment_id": self.experiment_id,
            "num_rounds": len(rounds),
            "final_train_loss": train_losses[-1] if train_losses else None,
            "final_ndcg@10": ndcg_scores[-1] if ndcg_scores else None,
            "final_hit@10": hit_scores[-1] if hit_scores else None,
            "best_ndcg@10": max(ndcg_scores) if ndcg_scores else None,
            "best_hit@10": max(hit_scores) if hit_scores else None,
            "config": self.experiment_data["config"]
        }


def create_experiment_id(
    dp_epsilon: Optional[float] = None,
    alpha: Optional[float] = None,
    embedding_dim: int = 16,
    num_clients: int = 3,
    seed: int = 42
) -> str:
    """
    Create standardized experiment ID.
    
    Args:
        dp_epsilon: DP budget (None for no DP)
        alpha: Heterogeneity parameter
        embedding_dim: Model embedding dimension
        num_clients: Number of clients
        seed: Random seed
    
    Returns:
        Experiment ID string
    """
    parts = []
    
    if dp_epsilon is None:
        parts.append("dp_inf")
    else:
        parts.append(f"dp_{dp_epsilon}")
    
    if alpha is not None:
        parts.append(f"alpha_{alpha}")
    
    parts.append(f"dim_{embedding_dim}")
    parts.append(f"clients_{num_clients}")
    parts.append(f"seed_{seed}")
    
    return "_".join(parts)

File: federated_learning_in_mobile/lib/utils/metrics_collector.dart
import 'dart:convert';
import 'dart:io';
import 'package:path_provider/path_provider.dart';
import 'package:csv/csv.dart';
import 'package:flutter/foundation.dart';
import 'package:path/path.dart' as path;

/// Metrics Collector for Federated Learning Experiments (Mobile)
/// Collects and saves all experimental metrics to JSON/CSV
class MetricsCollector {
  final String experimentId;
  late Directory resultsDir;
  
  // Store all collected metrics
  final Map<String, dynamic> experimentData;
  
  MetricsCollector({
    required this.experimentId,
  }) : experimentData = {
          'experiment_id': experimentId,
          'timestamp': DateTime.now().toIso8601String(),
          'config': <String, dynamic>{},
          'rounds': <Map<String, dynamic>>[],
          'final_metrics': <String, dynamic>{},
          'client_metrics': <Map<String, dynamic>>[],
          'mobile_metrics': <Map<String, dynamic>>[],
        };
  
  /// Initialize the results directory - saves to Downloads folder for easy access
  Future<void> initialize() async {
    // Check if running on web (not supported)
    try {
      if (kIsWeb) {
        throw UnsupportedError('Metrics collection not supported on web');
      }
    } catch (e) {
      // kIsWeb might not be available, continue anyway
    }
    
    // Try to use external storage Downloads folder (most accessible)
    try {
      final externalDir = await getExternalStorageDirectory();
      if (externalDir != null) {
        // Navigate to Downloads folder
        // Android external storage structure: /storage/emulated/0/Download
        final downloadsPath = Platform.isAndroid
            ? '/storage/emulated/0/Download/FL_Results'
            : '${externalDir.path}/Downloads/FL_Results';
        
        // Try the standard Downloads path first
        final downloadsDir = Directory(downloadsPath);
        
        // If that doesn't exist, try alternative paths
        if (!await downloadsDir.exists() && Platform.isAndroid) {
          // Alternative: Use external storage root + Download
          final altPath = path.join(externalDir.parent.path, 'Download', 'FL_Results');
          resultsDir = Directory(altPath);
        } else {
          resultsDir = downloadsDir;
        }
        
        if (!await resultsDir.exists()) {
          await resultsDir.create(recursive: true);
        }
        
        print('[METRICS_COLLECTOR] Using Downloads folder: ${resultsDir.path}');
        return;
      }
    } catch (e) {
      print('[METRICS_COLLECTOR] Warning: Downloads folder failed: $e');
    }
    
    // Fallback: Try external storage directory
    try {
      final externalDir = await getExternalStorageDirectory();
      if (externalDir != null) {
        resultsDir = Directory('${externalDir.path}/FL_Results');
        if (!await resultsDir.exists()) {
          await resultsDir.create(recursive: true);
        }
        print('[METRICS_COLLECTOR] Using external storage: ${resultsDir.path}');
        return;
      }
    } catch (e) {
      print('[METRICS_COLLECTOR] Warning: External storage failed: $e');
    }
    
    // Fallback: Application documents directory
    try {
      final appDocDir = await getApplicationDocumentsDirectory();
      resultsDir = Directory('${appDocDir.path}/FL_Results');
      
      if (!await resultsDir.exists()) {
        await resultsDir.create(recursive: true);
      }
      
      print('[METRICS_COLLECTOR] Using app documents: ${resultsDir.path}');
      return;
    } catch (e) {
      print('[METRICS_COLLECTOR] Warning: Application documents directory failed: $e');
    }
    
    // Last resort: temporary directory
    try {
      final tempDir = await getTemporaryDirectory();
      resultsDir = Directory('${tempDir.path}/FL_Results');
      if (!await resultsDir.exists()) {
        await resultsDir.create(recursive: true);
      }
      print('[METRICS_COLLECTOR] Using temporary directory: ${resultsDir.path}');
    } catch (e) {
      throw Exception('Could not initialize any storage directory: $e');
    }
  }
  
  /// Get the results directory path (for display)
  String get resultsPath => resultsDir.path;
  
  /// Get list of all result files
  Future<List<File>> getResultFiles() async {
    try {
      if (!await resultsDir.exists()) {
        return [];
      }
      
      final files = resultsDir.listSync()
          .whereType<File>()
          .where((f) => f.path.endsWith('.json') || f.path.endsWith('.csv'))
          .toList();
      
      return files;
    } catch (e) {
      print('[METRICS_COLLECTOR] Error listing files: $e');
      return [];
    }
  }
  
  /// Set experiment configuration
  void setConfig(Map<String, dynamic> config) {
    experimentData['config'] = config;
    print('[METRICS_COLLECTOR] Set experiment config');
  }
  
  /// Add metrics for a training round
  void addRoundMetrics({
    required int roundNum,
    required double trainLoss,
    Map<String, dynamic>? testMetrics,
    Map<String, dynamic>? aggregationInfo,
    List<Map<String, dynamic>>? clientMetrics,
    Map<String, dynamic>? resourceMetrics,
  }) {
    final roundData = <String, dynamic>{
      'round': roundNum,
      'train_loss': trainLoss,
      'test_metrics': testMetrics ?? <String, dynamic>{},
      'aggregation': aggregationInfo ?? <String, dynamic>{},
      'client_metrics': clientMetrics ?? <Map<String, dynamic>>[],
      'resource_metrics': resourceMetrics ?? <String, dynamic>{},
      'timestamp': DateTime.now().toIso8601String(),
    };
    
    (experimentData['rounds'] as List).add(roundData);
    print('[METRICS_COLLECTOR] Added metrics for round $roundNum');
  }
  
  /// Add final evaluation metrics
  void addFinalMetrics(Map<String, dynamic> metrics) {
    experimentData['final_metrics'] = metrics;
    print('[METRICS_COLLECTOR] Added final metrics');
  }
  
  /// Add per-client metrics
  void addClientMetrics(String clientId, Map<String, dynamic> metrics) {
    final clientData = <String, dynamic>{
      'client_id': clientId,
      ...metrics,
      'timestamp': DateTime.now().toIso8601String(),
    };
    
    (experimentData['client_metrics'] as List).add(clientData);
  }
  
  /// Add mobile device resource metrics
  void addMobileMetrics(String deviceId, Map<String, dynamic> metrics) {
    final mobileData = <String, dynamic>{
      'device_id': deviceId,
      ...metrics,
      'timestamp': DateTime.now().toIso8601String(),
    };
    
    (experimentData['mobile_metrics'] as List).add(mobileData);
  }
  
  /// Save metrics to JSON file
  Future<String> saveJson([String? filename]) async {
    await initialize();
    
    if (filename == null) {
      filename = '$experimentId.json';
    }
    
    final file = File('${resultsDir.path}/$filename');
    
    // Convert to JSON with proper formatting
    final jsonString = JsonEncoder.withIndent('  ').convert(experimentData);
    await file.writeAsString(jsonString);
    
    print('[METRICS_COLLECTOR] Saved JSON to: ${file.path}');
    return file.path;
  }
  
  /// Save summary to CSV file
  Future<String> saveCsvSummary([String? filename]) async {
    await initialize();
    
    if (filename == null) {
      filename = '${experimentId}_summary.csv';
    }
    
    final file = File('${resultsDir.path}/$filename');
    
    // Create CSV rows
    final List<List<dynamic>> csvData = [];
    
    // Header
    csvData.add([
      'Round',
      'Train_Loss',
      'NDCG@10',
      'Hit@10',
      'Precision@10',
      'Recall@10',
      'MSE',
      'MAE',
      'Accuracy',
      'Num_Clients',
      'Total_Samples',
      'Training_Time_MS',
      'Battery_Drain',
    ]);
    
    // Round data
    final rounds = experimentData['rounds'] as List;
    for (final roundData in rounds) {
      final testMetrics = roundData['test_metrics'] as Map<String, dynamic>? ?? {};
      final agg = roundData['aggregation'] as Map<String, dynamic>? ?? {};
      final resourceMetrics = roundData['resource_metrics'] as Map<String, dynamic>? ?? {};
      
      csvData.add([
        roundData['round'],
        roundData['train_loss']?.toString() ?? '',
        testMetrics['NDCG@10']?.toString() ?? '',
        testMetrics['Hit@10']?.toString() ?? '',
        testMetrics['Precision@10']?.toString() ?? '',
        testMetrics['Recall@10']?.toString() ?? '',
        testMetrics['mse']?.toString() ?? '',
        testMetrics['mae']?.toString() ?? '',
        testMetrics['accuracy']?.toString() ?? '',
        agg['num_clients']?.toString() ?? '',
        agg['total_samples']?.toString() ?? '',
        resourceMetrics['training_time_ms']?.toString() ?? '',
        resourceMetrics['battery_drain']?.toString() ?? '',
      ]);
    }
    
    // Final metrics row
    final finalMetrics = experimentData['final_metrics'] as Map<String, dynamic>? ?? {};
    csvData.add([
      'FINAL',
      '',
      finalMetrics['NDCG@10']?.toString() ?? '',
      finalMetrics['Hit@10']?.toString() ?? '',
      finalMetrics['Precision@10']?.toString() ?? '',
      finalMetrics['Recall@10']?.toString() ?? '',
      finalMetrics['mse']?.toString() ?? '',
      finalMetrics['mae']?.toString() ?? '',
      finalMetrics['accuracy']?.toString() ?? '',
      '',
      '',
      '',
      '',
    ]);
    
    // Write CSV
    final csvString = const ListToCsvConverter().convert(csvData);
    await file.writeAsString(csvString);
    
    print('[METRICS_COLLECTOR] Saved CSV summary to: ${file.path}');
    return file.path;
  }
  
  /// Get summary statistics
  Map<String, dynamic> getSummary() {
    final rounds = experimentData['rounds'] as List;
    
    if (rounds.isEmpty) {
      return {'message': 'No rounds collected yet'};
    }
    
    // Extract metrics across rounds
    final trainLosses = rounds
        .map((r) => (r as Map)['train_loss'] as double?)
        .where((loss) => loss != null)
        .cast<double>()
        .toList();
    
    final ndcgScores = rounds
        .map((r) {
          final testMetrics = (r as Map)['test_metrics'] as Map?;
          return testMetrics?['NDCG@10'] as double?;
        })
        .where((score) => score != null)
        .cast<double>()
        .toList();
    
    final hitScores = rounds
        .map((r) {
          final testMetrics = (r as Map)['test_metrics'] as Map?;
          return testMetrics?['Hit@10'] as double?;
        })
        .where((score) => score != null)
        .cast<double>()
        .toList();
    
    return {
      'experiment_id': experimentId,
      'num_rounds': rounds.length,
      'final_train_loss': trainLosses.isNotEmpty ? trainLosses.last : null,
      'final_ndcg@10': ndcgScores.isNotEmpty ? ndcgScores.last : null,
      'final_hit@10': hitScores.isNotEmpty ? hitScores.last : null,
      'best_ndcg@10': ndcgScores.isNotEmpty ? ndcgScores.reduce((a, b) => a > b ? a : b) : null,
      'best_hit@10': hitScores.isNotEmpty ? hitScores.reduce((a, b) => a > b ? a : b) : null,
      'config': experimentData['config'],
    };
  }
}

/// Create standardized experiment ID
String createExperimentId({
  double? dpEpsilon,
  double? alpha,
  int embeddingDim = 16,
  int numClients = 3,
  int? seed,
  String? deviceId,
}) {
  final parts = <String>[];
  
  if (dpEpsilon == null) {
    parts.add('dp_inf');
  } else {
    parts.add('dp_$dpEpsilon');
  }
  
  if (alpha != null) {
    parts.add('alpha_$alpha');
  }
  
  parts.add('dim_$embeddingDim');
  parts.add('clients_$numClients');
  
  if (seed != null) {
    parts.add('seed_$seed');
  }
  
  if (deviceId != null) {
    // Clean device ID for filename
    final cleanId = deviceId.replaceAll(RegExp(r'[^\w-]'), '_');
    parts.add('device_$cleanId');
  }
  
  // Add timestamp for uniqueness
  final timestamp = DateTime.now().millisecondsSinceEpoch;
  parts.add('t$timestamp');
  
  return parts.join('_');
}

